### DISCUSSION TITLE
SchedulerContext can hold data of any type but JobDataMap  can not

### DISCUSSION THREAD

**Discussion by reponemec** [quartznet/quartznet#1080] [Q&A] [answered]
Why **SchedulerContext** can hold data of any type (e.g. IFileScanListener)...

```
//This is from FileScanJob.Execute:
public virtual async Task Execute(IJobExecutionContext context)
...
SchedulerContext schedCtxt = context.Scheduler.Context;
IFileScanListener listener = (IFileScanListener) schedCtxt[listenerName];
...
```

...while **JobDataMap** can only hold data of primitive types?

**Comment by lahma** [comment, marked as answer]
`JobDataMap` can also hold objects?

```csharp
IJobDetail job = JobBuilder.Create<MyJob>().WithIdentity("job", "group").Build();
job.JobDataMap.Put("key", new object());
job.JobDataMap["key2"] = new object();
````

**Comment by reponemec** [comment]
Yes, technically speaking, it is possible. However Best practices say:

_"Only store primitive data types (including strings) in JobDataMap to avoid data serialization issues short and long-term."_ 
https://www.quartz-scheduler.net/documentation/best-practices.html#jobdatamap-tips.

Why that practice is not relevant for SchedulerContext? In other words, SchedulerContext is not subject of serialization?

**Comment by lahma** [comment]
Scheduler context is not serialized so it isn't vulnerable to object format changes. If you are using database to store data primitive types + JSON serializer is always a wise choice.

**Reply by reponemec** [reply]
For example, if **FileScanJob's** trigger misfire its execution (suppose scheduler itself was down) and later will be reexecuted by the scheduler, that will throw JobExecutionException?

https://github.com/quartznet/quartznet/blob/a22915a9abac1568accb93eb24b4cce5331c8249/src/Quartz.Jobs/Job/FileScanJob.cs#L126

**Reply by lahma** [reply]
That would be quite odd situation. The listener is set as part of starting the scheduler (like in `XMLSchedulingDataProcessorPlugin.Start`), it's part of the scheduler setup and not part of the job / trigger state per se. Have you actually encountered a problem with this?

**Reply by reponemec** [reply]
_...it's part of the **scheduler** setup and not part of the job / trigger._
It's clear to me now, thanks.

**Comment by asulwer** [comment]
did i misunderstand what is being said here?  this line of code is not proper?

var jobMap = new JobDataMap();
jobMap.Put("Site", new Site());

**Reply by lahma** [reply]
It's proper. You just need to be aware of the versioning issues that might rise if Site's structure changes.
